### PR TITLE
SAMZA-1221, SAMZA-1101: Internal cleanup for High-Level API implementation.

### PR DESCRIPTION
**Pull request by prateekm**
SAMZA-1221: Separated the OperatorSpec and MessageStream DAGs so that they're now duals of each other. Users interact with and construct the MessageStream DAG; we create and use the OperatorSpec DAG internally.
Moved common OperatorSpec functionality (getId, getOpCode, getOpName etc.) to the OperatorSpec abstract base class.
Added a new JoinOperatorSpec and PartialJoinOperatorImpls which are created from JoinOperatorSpec in OperatorGraphImpl.
Added a new InputOperatorSpec and InputOperatorImpl (previously RootOperatorImpl). InputOperatorSpec is created when StreamGraph#getInputStream is called.
SAMZA-1101: Added a new OutputOperatorSpec and OutputOperatorImpl for partitionBy and sendTo. These are Separate from SinkOperatorSpec for and SinkOperatorImpl for sink. We don't need to create a sinkFn for partitionBy and sendTo anymore.
Updated most unit tests to use the new classes and avoid reflection.

### CODE REVIEW

**Comment by prateekm** [comment]
@nickpan47 @vjagadish1989 @xinyuiscool, please take a look. Probably best to check out and view changes in an IDE.

**Comment by prateekm** [comment]
+ @bharathkk too.

File: samza-core/src/main/java/org/apache/samza/execution/ExecutionPlanner.java [hunk 1]
```diff
@@ -210,43 +204,31 @@ public ExecutionPlan plan(StreamGraphImpl streamGraph) throws Exception {
 
   /**
    * This function traverses the StreamGraph to find and update mappings for all Joins reachable from this input StreamEdge
-   * @param inputMessageStream next input MessageStream to traverse {@link MessageStream}
+   * @param operatorSpec the {@link OperatorSpec} to traverse
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/execution/ExecutionPlanner.java]: Updated, thanks.

File: samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java [hunk 2]
```diff
@@ -38,127 +43,99 @@
...
 
-  public OperatorImplGraph(Clock clock) {
-    this.clock = clock;
-  }
+  /**
+   * A mapping from {@link JoinOperatorSpec}s to their two {@link PartialJoinFunction}s. Used to associate
+   * the two {@link PartialJoinOperatorImpl}s for a {@link JoinOperatorSpec} with each other since they're
+   * reached from different {@link OperatorSpec} during DAG traversals.
+   */
+  private final Map<Integer, Pair<PartialJoinFunction, PartialJoinFunction>> joinFunctions = new HashMap<>();
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java]: It's required for correctness. `computeIfAbsent` returns current value if present. We create both partial join implementations the first time we encounter the join operator spec. The next time we see this we get and use the other side.

File: samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java [hunk 3]
```diff
@@ -167,20 +144,90 @@ public RootOperatorImpl getRootOperator(SystemStream systemStream) {
...
         String.format("Unsupported OperatorSpec: %s", operatorSpec.getClass().getName()));
   }
+
+  private PartialJoinOperatorImpl createPartialJoinOperatorImpl(OperatorSpec prevOperatorSpec,
+      JoinOperatorSpec joinOpSpec, Config config, TaskContext context, Clock clock) {
+    Pair<PartialJoinFunction, PartialJoinFunction> joinFunctions = getOrCreatePartialJoinFunctions(joinOpSpec);
+
+    if (joinOpSpec.getLeftInputOpSpec().equals(prevOperatorSpec)) { // we got here from the left side of the join
+      return new PartialJoinOperatorImpl(joinOpSpec, /* isLeftSide */ true,
+          joinFunctions.getLeft(), joinFunctions.getRight(), config, context, clock);
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java]: Good point, renamed to partialJoinFunctions

File: samza-core/src/main/java/org/apache/samza/execution/JobGraphJsonGenerator.java [hunk 4]
```diff
@@ -73,8 +70,6 @@
     List<StreamJson> outputStreams;
     @JsonProperty("operators")
     Map<Integer, Map<String, Object>> operators = new HashMap<>();
-    @JsonProperty("canonicalOpIds")
-    Map<Integer, String> canonicalOpIds = new HashMap<>();
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/execution/JobGraphJsonGenerator.java]: Yeah AFAIK this is not needed anymore. @xinyuiscool can confirm.

File: samza-core/src/main/java/org/apache/samza/operators/spec/OperatorSpec.java [hunk 5]
```diff
@@ -41,38 +45,72 @@
...
-   * @return  the next {@link MessageStreamImpl}
+   * Register the next operator spec in the chain that this operator should propagate its output to.
+   * @param nextOperatorSpec  the next operator in the chain.
    */
-  MessageStreamImpl<OM> getNextStream();
+  public void registerNextOperatorSpec(OperatorSpec<OM, ?> nextOperatorSpec) {
+    nextOperatorSpecs.add(nextOperatorSpec);
+  }
+
+  public Collection<OperatorSpec<OM, ?>> getRegisteredOperatorSpecs() {
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/operators/spec/OperatorSpec.java]: They were non-final for testing (can't mock final methods using mockito).

**Comment by prateekm** [comment]
@vjagadish1989 please take a look again.

File: samza-core/src/main/java/org/apache/samza/operators/spec/OperatorSpec.java [hunk 6]
```diff
@@ -41,39 +45,77 @@
...
+
+  /**
+   * Register the next operator spec in the chain that this operator should propagate its output to.
+   * @param nextOperatorSpec  the next operator in the chain.
+   */
+  public void registerNextOperatorSpec(OperatorSpec<OM, ?> nextOperatorSpec) {
+    nextOperatorSpecs.add(nextOperatorSpec);
+  }
+
+  public Collection<OperatorSpec<OM, ?>> getRegisteredOperatorSpecs() {
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/operators/spec/OperatorSpec.java]: Can we change the name to getNextOperatorSpecs()?

**Pull request closed without merging**